formats/json-tests/jvmTest/src/kotlinx/serialization/JavaCollectionsTest.kt

```diff
@@ -49,4 +51,55 @@ class JavaCollectionsTest {
         val restored = Json.decodeFromString(deserializer = serializer, string = string)
         assertEquals(expected = original, actual = restored)
     }
+
+    @Test
+    fun testTopLevelMaps() {
+        // Returning null here is a deliberate choice: map constructor functions may return different specialized
+        // implementations (e.g., kotlin.collections.EmptyMap or java.util.Collections.SingletonMap)
+        // that may or may not be generic. Since we generally cannot return a generic serializer using Java class only,
+        // all attempts to get map serializer using only .javaClass should return null.
+        assertNull(serializerOrNull(emptyMap<String, String>().javaClass))
```

Comment:
There are two additional tests that should be added here. This adds two additional test cases (both fail with the suggested implementation):
- Getting a serializer from  an anonymous object (`canonicalName` returns `null`) throws an NPE
- Getting a serializer from a local named object throws because it is not built in

```suggestion
    private object namedObject

    @Test
    fun testTopLevelMaps() {
        // Returning null here is a deliberate choice: map constructor functions may return different specialized
        // implementations (e.g., kotlin.collections.EmptyMap or java.util.Collections.SingletonMap)
        // that may or may not be generic. Since we generally cannot return a generic serializer using Java class only,
        // all attempts to get map serializer using only .javaClass should return null.
        assertNull(serializerOrNull(emptyMap<String, String>().javaClass))
        assertNull(serializerOrNull(object {}.javaClass))
        assertNull(serializerOrNull(namedObject.javaClass))
```

Reply:
Thanks for pointing out the problem with the anonymous class! Adding a safe call to `canonicalName` is enough to fix it, though.